implementations/rust/ockam/ockam_api/src/authenticator/storage/authority_members_repository_sql.rs

```diff
@@ -0,0 +1,241 @@
+use sqlx::*;
+use tracing::debug;
+
+use ockam::identity::Identifier;
+use ockam_core::async_trait;
+use ockam_core::Result;
+use ockam_node::database::{FromSqlxError, SqlxDatabase, ToSqlxType, ToVoid};
+
+use crate::authenticator::{
+    AuthorityMember, AuthorityMemberRow, AuthorityMembersRepository, PreTrustedIdentities,
+};
+
+#[derive(Clone)]
+pub struct AuthorityMembersSqlxDatabase {
+    database: SqlxDatabase,
+}
+
+impl AuthorityMembersSqlxDatabase {
+    /// Create a new database
+    pub fn new(database: SqlxDatabase) -> Self {
+        debug!("create a repository for authority members");
+        Self { database }
+    }
+
+    /// Create a new in-memory database
+    pub async fn create() -> Result<Self> {
+        Ok(Self::new(
+            SqlxDatabase::in_memory("authority members").await?,
+        ))
+    }
+}
+
+#[async_trait]
+impl AuthorityMembersRepository for AuthorityMembersSqlxDatabase {
+    async fn get_member(&self, identifier: &Identifier) -> Result<Option<AuthorityMember>> {
+        let query = query_as("SELECT identifier, attributes, added_by, added_at, is_pre_trusted FROM authority_member WHERE identifier=?")
+            .bind(identifier.to_sql());
+        let row: Option<AuthorityMemberRow> = query
+            .fetch_optional(&*self.database.pool)
+            .await
+            .into_core()?;
+        row.map(|r| r.try_into()).transpose()
+    }
+
+    async fn get_members(&self) -> Result<Vec<AuthorityMember>> {
+        let query = query_as("SELECT identifier, attributes, added_by, added_at, is_pre_trusted FROM authority_member");
+        let row: Vec<AuthorityMemberRow> =
+            query.fetch_all(&*self.database.pool).await.into_core()?;
+        row.into_iter().map(|r| r.try_into()).collect()
+    }
+
+    async fn delete_member(&self, identifier: &Identifier) -> Result<()> {
+        let query = query("DELETE FROM authority_member WHERE identifier=? AND is_pre_trusted=?")
+            .bind(identifier.to_sql())
+            .bind(false.to_sql());
+        query.execute(&*self.database.pool).await.void()
+    }
+
+    async fn add_member(&self, member: AuthorityMember) -> Result<()> {
+        let query = query("INSERT OR REPLACE INTO authority_member VALUES (?1, ?2, ?3, ?4, ?5)")
+            .bind(member.identifier().to_sql())
+            .bind(member.added_by().clone().map(|x| x.to_sql()))
+            .bind(member.added_at().to_sql())
+            .bind(member.is_pre_trusted().to_sql())
+            .bind(minicbor::to_vec(member.attributes())?.to_sql());
+
+        query.execute(&*self.database.pool).await.void()
+    }
+
+    async fn bootstrap_pre_trusted_members(
```

Comment:
I don't see a good reason for not having this loop outside the database code, inside `authority.rs`.
This way we separate the general notion of "member", from the fact that there are identities that we pre-trust.

Reply:
Or maybe we should go the other way round? Make sure that pretrused members can only be added here? Make `add_member` reject such members. Thoughts?